### ISSUE TITLE
Redux-Form <Field> with a custom component TS 3.2 

### ISSUE THREAD

**Issue by virzak** [DefinitelyTyped/DefinitelyTyped#31215]
- Authors: @Kovensky @carsonf @aikoven @LKay @bancek @Aleksion @weswigham

Seems like none of the test cases has a `<Field>` with a custom component, even though it is everywhere in the examples ([here](https://github.com/erikras/redux-form/blob/734be8db5aaf5316f2c935ac343971a5e4f8ec8f/examples/material-ui/src/MaterialUiForm.js#L94)). It also fails when switching to TS 3.2, but works with TS 3.1

Is it a must to use `<FieldCustom>` as per the [test](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/redux-form/redux-form-tests.tsx#L136)?

If so why wasn't this a problem before 3.2?

**Comment by LKay** [comment]
TS is not able to resolve the correct outer type for a field when a custom component (at least as of <3.1) on its own and thus the type overload using: `const FieldCustom = Field as new () => GenericField<MyFieldCustomProps>;` as defined in the tests. Not sure if something has changed in the latest version of TS (which seems unlikely, afaik). I would need to investigate myself why it's not working in TS3.2 as it was before.

**Comment by virzak** [comment]
@LKay `const FieldCustom = Field as new () => GenericField<MyFieldCustomProps>;` works perfectly fine with TS 3.2, but `<Field>` doesn't. `<Field>` was working fine with TS 3.1 without any overloads for me. If you're saying that overload should be used regardless, perhaps this is not an issue at all. However, usually, when converting from JS to TS, the same tags can be used.

**Comment by sargant** [comment]
I'm also getting type failures using `@types/redux-form@7.4.12` and `typescript@3.2.2` when trying to use a custom component with a `<Field>`, which where not failures using `typescript@3.1.6`.

Minimal code that fails:
```js
import { FC } from 'react'
import { Field } from 'redux-form'
const MyInput: FC = () => <input type='text' />
const MyForm: FC = () => <Field name='foo' component={MyInput} />
```

**Comment by Razinsky** [comment]
Same thing here, on `typescript@3.2.2`

`Type 'typeof 'CustomTextField' is not assignable to type '"input"'.`

For now I can silence it by doing that but that's bad :)
```
<Field<{}> component={CustomTextField} name="textfield" />
```

**Comment by virzak** [comment]
@Razinsky `<Field<{}>` doesn't work either

**Comment by jimjeffers** [comment]
I am having the same problem. Will stay tuned to see what the solution ends up being. For now I am casting as any during assignment which is not ideal...

```
<Field name="fullName" component={Form.Input as any} type="text" />
```

**Comment by breeeew** [comment]
> Has anyone considered redefining `Field`

yes, that seems works

**Comment by sargant** [comment]
Fixed in #32854 @virzak

**Comment by virzak** [comment]
Thanks a bunch!
